<a href="https://colab.research.google.com/github/jesusvasquez-tec/analitica-empresarial-integrada/blob/main/Semana3/Vasquez_LAB_D3_AEI_PSAYAN_2026_02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# LABORATORIO N.° 03 — La métrica que importa

**Curso:** Analítica Empresarial Integrada  
**Semana 3:** KPI accionables, North Star Metric y árbol de métricas  
**Docente:** Pilar Rocío Sayán Mejía  
**Periodo:** 2026-II

**Fuente real:** UCI Machine Learning Repository — Online Retail (ID 352).

> Objetivo del notebook: conectar **objetivo → North Star → drivers → guardrails → KPI → tablero → interpretación → decisión** usando transacciones reales.
---

**Equipo N.°:** 03 &nbsp;&nbsp;&nbsp; **Sección:** 6C28A &nbsp;&nbsp;&nbsp; **Fecha:** 02-09-2026

**Apellidos y nombres del estudiante:** Vasquez Espinoza, Jesus

**Integrantes del equipo:**

* Alvino Carhuaricra, Enzo
* Cordova Agurto, Isabel
* Palomino Acuña, Luis
* Vasquez Espinoza, Jesus

---


## Agenda de laboratorio — 7:00 p. m. a 10:10 p. m.

**Duración total:** 190 minutos · **Receso:** 15 minutos · **Trabajo efectivo:** 175 minutos.

| Horario | Tiempo | Desarrollo |
|---|---:|---|
| 7:00–7:10 | 10 min | Apertura del caso y presentación del problema de medición. |
| 7:10–7:35 | 25 min | Actividad 1. Revisión de conceptos: del dato a la decisión. |
| 7:35–8:00 | 25 min | Actividad 2. Descarga desde UCI, auditoría y limpieza documentada. |
| 8:00–8:30 | 30 min | Actividad 2. Periodo comparable, recurrencia y North Star. Reto 1. |
| 8:30–8:45 | 15 min | **RECESO** |
| 8:45–9:15 | 30 min | Actividad 2. Árbol de métricas, guardrails, Polars y DuckDB. |
| 9:15–9:35 | 20 min | Actividad 2. Tablero de decisión en Plotly. Reto 2. |
| 9:35–10:00 | 25 min | **Reto de aplicación y retroalimentación.** Ejercicios 1 a 5. |
| 10:00–10:10 | 10 min | Diccionario de KPI, informe ejecutivo y ticket de salida. |

> **Regla de trabajo:** no avance de bloque sin registrar la interpretación solicitada. El objetivo no es ejecutar celdas, sino convertir datos en evidencia para una decisión.


## Actividad 1 — Revisión de conceptos: del dato a la decisión (25 minutos)

**Propósito.** Establecer con precisión el vocabulario de medición antes de programar. La confusión entre dato, métrica, indicador y KPI constituye la causa más frecuente de tableros que no sustentan ninguna decisión.

**Instrucciones.** Complete la tabla con definiciones elaboradas con sus propias palabras. No se admite la reproducción literal de fuentes externas ni de sistemas generativos. La columna de la derecha contiene una pregunta de apoyo: si su definición permite responderla, la definición es suficiente; si no lo permite, corríjala antes de continuar.

**Evidencia esperada.** Tabla completa con las definiciones registradas.

| Concepto | Definición elaborada por el estudiante | Pregunta de apoyo |
|---|---|---|
| Dato | Registro elemental y sin procesar que describe un hecho o transacción específica, <br>carente de contexto analítico por sí solo. | ¿En qué se diferencia un dato de una métrica? Proponga un ejemplo de la base utilizada. |
| Métrica | Medida cuantitativa calculada a partir de la agregación o relación de datos crudos, <br>la cual por sí sola no indica si el desempeño es bueno o malo. | ¿Toda métrica calculada correctamente resulta útil para decidir? Justifique. |
| Indicador | Métrica a la que se le añade un contexto, un punto de referencia o un umbral de <br>comparación que permite evaluar el desempeño de manera relativa. | ¿Qué debe añadirse a una métrica para que constituya un indicador? |
| KPI | Indicador clave vinculado directamente a un objetivo estratégico crítico, que cuenta <br>con un responsable y una acción asociada para su mejora. | ¿Por qué una organización no puede sostener veinte KPI simultáneos? |
| Meta | Valor objetivo o umbral de desempeño futuro que la organización o el equipo se <br>propone alcanzar en un periodo determinado. | ¿Qué distingue un valor observado en la base de una meta propuesta para el ejercicio? |
| North Star | Métrica principal que representa el valor recurrente y sostenible que el negocio <br>entrega a sus clientes, conectando su experiencia con el crecimiento empresarial. | ¿Qué condición debe cumplir una North Star para no convertirse en métrica de vanidad? |
| Driver | Factor operativo o variable de primer nivel que impulsa, explica matemáticamente <br>o descompone el comportamiento de la North Star. | ¿Cómo se comprueba que una variable es efectivamente driver de la North Star? |
| Guardrail | Métrica de seguridad que establece límites para proteger la calidad, el margen o <br>la sostenibilidad del negocio mientras se optimiza la North Star. | ¿Qué ocurre si una organización optimiza su North Star sin vigilar los guardrails? |

**Criterio de cierre.** No se avanza al desarrollo práctico mientras existan conceptos sin definición registrada.


## Actividad 2 — Desarrollo práctico y ejecución

### Presentación del caso

Una empresa minorista en línea del Reino Unido cuenta con un registro histórico de transacciones que incluye facturas, productos, cantidades, fechas, precios unitarios, clientes y países. La dirección necesita transformar estos registros en un sistema de métricas que permita distinguir crecimiento útil de crecimiento aparente. Para ello, no basta con observar ventas acumuladas o cantidad de clientes: es necesario identificar qué indicadores representan valor recurrente para el cliente y cuáles pueden orientar una decisión empresarial.

Durante el laboratorio, el equipo trabajará con el conjunto Online Retail del UCI Machine Learning Repository. A partir de los datos reales, deberá auditar y preparar las transacciones, identificar clientes recurrentes, proponer y justificar una North Star, descomponerla en drivers y guardrails, construir un diccionario de KPI y elaborar un tablero de decisión en Plotly. El análisis deberá terminar con hallazgos cuantitativos y una acción empresarial concreta, diferenciando en todo momento los valores observados en la base de las metas o umbrales académicos propuestos para el ejercicio.


## Paso 1 — Preparación reproducible del entorno


In [ ]:
%pip install -q ucimlrepo==0.0.7 polars==1.17.1 duckdb==1.1.3

import polars as pl
import duckdb
import pandas as pd            # solo para recibir la descarga de UCI y alimentar Plotly Express
import plotly.express as px
import plotly.graph_objects as go
from IPython.display import display
from ucimlrepo import fetch_ucirepo

pl.Config.set_tbl_rows(20)
pl.Config.set_tbl_width_chars(160)

# Formato REAL de las fechas de esta fuente: 12/1/2010 8:26 -> mes/dia/anio hora:min.
# Se declara de forma explicita en lugar de dejar que la libreria lo adivine.
FORMATO_FECHA = "%m/%d/%Y %H:%M"

print("polars:", pl.__version__, "| duckdb:", duckdb.__version__)
print("Entorno listo.")


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 36.4/36.4 MB 12.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.2/20.2 MB 18.6 MB/s eta 0:00:00
polars: 1.17.1 | duckdb: 1.1.3
Entorno listo.


## Paso 2 — Descarga de datos reales desde UCI

El conjunto **Online Retail** contiene transacciones de una empresa minorista en línea registrada en el Reino Unido entre diciembre de 2010 y diciembre de 2011. Los códigos de factura que empiezan con `C` representan cancelaciones. No se generan registros artificiales.


In [ ]:
online_retail = fetch_ucirepo(id=352)
df = pl.from_pandas(online_retail.data.original)   # la descarga llega en pandas; se pasa a Polars

print("Dataset:", online_retail.metadata.get("name"))
print("Filas y columnas:", df.shape)
print("Columnas:", df.columns)
display(df.head())


Dataset: Online Retail
Filas y columnas: (541909, 8)
Columnas: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']


InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
str,str,str,i64,str,f64,f64,str
"""536365""","""85123A""","""WHITE HANGING HEART T-LIGHT HO…",6,"""12/1/2010 8:26""",2.55,17850.0,"""United Kingdom"""
"""536365""","""71053""","""WHITE METAL LANTERN""",6,"""12/1/2010 8:26""",3.39,17850.0,"""United Kingdom"""
"""536365""","""84406B""","""CREAM CUPID HEARTS COAT HANGER""",8,"""12/1/2010 8:26""",2.75,17850.0,"""United Kingdom"""
"""536365""","""84029G""","""KNITTED UNION FLAG HOT WATER B…",6,"""12/1/2010 8:26""",3.39,17850.0,"""United Kingdom"""
"""536365""","""84029E""","""RED WOOLLY HOTTIE WHITE HEART.""",6,"""12/1/2010 8:26""",3.39,17850.0,"""United Kingdom"""


### Control de trazabilidad
Registre: número de filas, columnas, rango de fechas y países presentes.

**Respuesta:** El conjunto de datos original está conformado por 541,909 filas y 8 columnas. El registro de transacciones abarca un rango temporal desde el 1 de diciembre de 2010 hasta el 9 de diciembre de 2011. Geográficamente, la base documenta las operaciones de una empresa minorista del Reino Unido, incluyendo transacciones de clientes provenientes de 38 países distintos.



## Paso 3 — Auditoría inicial y limpieza documentada


In [ ]:
df = df.with_columns([
    pl.col("InvoiceNo").cast(pl.Utf8),
    pl.col("CustomerID").cast(pl.Utf8),
    pl.col("Quantity").cast(pl.Float64, strict=False),
    pl.col("UnitPrice").cast(pl.Float64, strict=False),
    pl.col("InvoiceDate").cast(pl.Utf8)
      .str.to_datetime(format=FORMATO_FECHA, strict=False).alias("InvoiceDate"),
])
df = df.with_columns([
    pl.col("InvoiceNo").str.to_uppercase().str.starts_with("C").alias("es_cancelacion"),
    (pl.col("Quantity") * pl.col("UnitPrice")).alias("importe_linea"),
])

# Control de parseo: si el formato declarado no fuese el correcto, aqui apareceria
# un conteo de fechas nulas. pandas habria "adivinado" un formato sin avisar;
# Polars obliga a declararlo y deja el error a la vista.
nulas = df["InvoiceDate"].null_count()
print("Fechas que no pudieron parsearse:", nulas)
if nulas:
    raise ValueError("El formato declarado en FORMATO_FECHA no corresponde a la fuente.")

# Control de integridad de la fuente. Si UCI republica el archivo, el laboratorio
# se detiene aqui en vez de producir numeros que no coinciden con el solucionario.
FILAS_ESPERADAS = 541_909
FACTURAS_ESPERADAS = 25_900

if df.height != FILAS_ESPERADAS:
    raise ValueError(
        f"La fuente cambio: se esperaban {FILAS_ESPERADAS:,} filas y llegaron {df.height:,}. "
        "Avise al docente antes de continuar."
    )
if df["InvoiceNo"].n_unique() != FACTURAS_ESPERADAS:
    raise ValueError(
        f"La fuente cambio: se esperaban {FACTURAS_ESPERADAS:,} facturas distintas "
        f"y llegaron {df['InvoiceNo'].n_unique():,}."
    )
print("Integridad verificada:", f"{df.height:,}", "filas y",
      f"{df['InvoiceNo'].n_unique():,}", "facturas, como se esperaba.")

control = pl.DataFrame({
    "indicador": ["filas", "facturas", "clientes", "cancelaciones",
                  "cantidades_no_positivas", "precios_no_positivos"],
    "valor": [df.height,
              df["InvoiceNo"].n_unique(),
              df["CustomerID"].drop_nulls().n_unique(),
              int(df["es_cancelacion"].sum()),
              int((df["Quantity"] <= 0).sum()),
              int((df["UnitPrice"] <= 0).sum())],
})
display(control)
print("Rango:", df["InvoiceDate"].min(), "->", df["InvoiceDate"].max())


Fechas que no pudieron parsearse: 0
Integridad verificada: 541,909 filas y 25,900 facturas, como se esperaba.


indicador,valor
str,i64
"""filas""",541909
"""facturas""",25900
"""clientes""",4372
"""cancelaciones""",9288
"""cantidades_no_positivas""",10624
"""precios_no_positivos""",2517


Rango: 2010-12-01 08:26:00 -> 2011-12-09 12:50:00


In [ ]:
compras = df.filter(
    (~pl.col("es_cancelacion"))
    & (pl.col("Quantity") > 0)
    & (pl.col("UnitPrice") > 0)
    & pl.col("InvoiceDate").is_not_null()
    & pl.col("CustomerID").is_not_null()
)

print("Filas originales:", df.height)
print("Filas de compra validas:", compras.height)
print("Proporcion conservada: {:.1%}".format(compras.height / df.height))


Filas originales: 541909
Filas de compra validas: 397884
Proporcion conservada: 73.4%


### Reto de calidad
1. ¿Qué sesgo aparece si contamos cancelaciones como ventas?  
2. ¿Por qué no debemos borrar esos registros de la fuente original?

**Respuesta:**  Contar las cancelaciones como ventas genera un sesgo de sobreestimación que infla artificialmente los ingresos y el volumen de transacciones, distorsionando el desempeño real del negocio. Por otro lado, no se deben eliminar estos registros de la fuente original porque constituyen eventos operativos legítimos que preservan la trazabilidad histórica de los datos; además, su conservación es indispensable para calcular métricas de control posteriores, como la tasa de cancelación, la cual se utiliza como un guardrail para evaluar la calidad y sostenibilidad del crecimiento.



## Paso 4 — Definición del periodo comparable

La fuente empieza el 1/12/2010 y termina el 9/12/2011. Para no comparar meses incompletos con meses completos, el laboratorio informa sobre **enero–noviembre de 2011**.

**Cuidado con el sesgo de ventana.** La tabla de facturas se construye sobre *todo* el historial disponible, y el recorte a la ventana se aplica **después** de determinar la recurrencia. Si se recorta antes, un cliente que compró en diciembre de 2010 y volvió en enero de 2011 aparece como comprador nuevo, y la North Star crece de forma artificial en los primeros meses: en esta base, enero pasa de 570 a 246 compras recurrentes, un 57 % menos, solo por haber filtrado en el orden equivocado.


In [ ]:
compras = compras.with_columns(pl.col("InvoiceDate").dt.strftime("%Y-%m").alias("mes"))
INICIO, FIN = pl.datetime(2011, 1, 1), pl.datetime(2011, 12, 1)

# La tabla se construye sobre TODO el historial: el recorte a la ventana
# comparable se hace mas adelante, una vez determinada la recurrencia.
facturas = (compras.group_by(["InvoiceNo", "CustomerID", "mes"])
            .agg([pl.col("InvoiceDate").min().alias("fecha_factura"),
                  pl.col("importe_linea").sum().alias("importe_factura"),
                  pl.col("Quantity").sum().alias("unidades"),
                  pl.col("StockCode").count().alias("lineas")]))

en_ventana = facturas.filter((pl.col("fecha_factura") >= INICIO) & (pl.col("fecha_factura") < FIN))
print("Facturas en el historial completo :", facturas["InvoiceNo"].n_unique())
print("Facturas en la ventana comparable :", en_ventana["InvoiceNo"].n_unique())
print("Clientes en la ventana            :", en_ventana["CustomerID"].n_unique())
print("Ingresos en la ventana (GBP)      :", round(en_ventana["importe_factura"].sum(), 2))
display(facturas.head())


Facturas en el historial completo : 18532
Facturas en la ventana comparable : 16354
Clientes en la ventana            : 4173
Ingresos en la ventana (GBP)      : 7820501.22


InvoiceNo,CustomerID,mes,fecha_factura,importe_factura,unidades,lineas
str,str,str,datetime[μs],f64,f64,u32
"""560724""","""16186.0""","""2011-07""",2011-07-20 13:55:00,534.34,464.0,29
"""536972""","""16926.0""","""2010-12""",2010-12-03 13:42:00,230.25,35.0,14
"""576379""","""16423.0""","""2011-11""",2011-11-15 09:21:00,346.12,230.0,18
"""549836""","""16814.0""","""2011-04""",2011-04-12 13:32:00,180.85,75.0,16
"""547937""","""15068.0""","""2011-03""",2011-03-28 14:27:00,239.96,130.0,39


## Paso 5 — Recurrencia y North Star

**Regla operativa del laboratorio:** un cliente se considera recurrente desde su segunda factura válida, contada sobre **todo el historial disponible**. Su primera compra no se reclasifica retrospectivamente.

Enero de 2011 conserva un sesgo residual, porque solo dispone de un mes previo de historial. Por eso se marca como **mes de calentamiento** y se excluye de las comparaciones de variación mensual.


In [ ]:
facturas = facturas.sort(["CustomerID", "fecha_factura", "InvoiceNo"])
facturas = facturas.with_columns(
    (pl.int_range(pl.len()).over("CustomerID") + 1).alias("n_compra_cliente"))

# Una factura es recurrente a partir de la SEGUNDA compra del cliente, en orden
# cronologico. La marca se calcula sobre el historial completo para no clasificar
# retroactivamente como recurrente la primera compra de un cliente que volvio despues.
facturas = facturas.with_columns(
    (pl.col("n_compra_cliente") >= 2).alias("es_recurrente"))

# Recien ahora se recorta a la ventana comparable: la recurrencia ya quedo
# determinada usando todo el historial, sin sesgo de ventana.
facturas = facturas.filter((pl.col("fecha_factura") >= INICIO) & (pl.col("fecha_factura") < FIN))

# Enero solo tiene un mes previo de historial: no es comparable en variacion.
MES_CALENTAMIENTO = "2011-01"

mensual = (facturas.group_by("mes")
           .agg([pl.col("InvoiceNo").n_unique().alias("facturas_validas"),
                 pl.col("CustomerID").n_unique().alias("clientes_activos"),
                 pl.col("importe_factura").sum().alias("ingresos")]))

rec = (facturas.filter("es_recurrente").group_by("mes")
       .agg([pl.col("InvoiceNo").n_unique().alias("compras_recurrentes"),
             pl.col("CustomerID").n_unique().alias("clientes_recurrentes"),
             pl.col("importe_factura").sum().alias("ingresos_recurrentes")]))

kpi = mensual.join(rec, on="mes", how="left").fill_null(0).sort("mes")
kpi = kpi.with_columns([
    (pl.col("compras_recurrentes") / pl.col("clientes_recurrentes")).alias("frecuencia_recurrente"),
    (100 * pl.col("ingresos_recurrentes") / pl.col("ingresos")).alias("participacion_ingreso_recurrente_pct"),
])
display(kpi)


mes,facturas_validas,clientes_activos,ingresos,compras_recurrentes,clientes_recurrentes,ingresos_recurrentes,frecuencia_recurrente,participacion_ingreso_recurrente_pct
str,u32,u32,f64,u32,u32,f64,f64,f64
"""2011-01""",987,741,569445.04,570,370,296614.01,1.540541,52.088259
"""2011-02""",997,758,447137.35,617,412,299937.36,1.497573,67.079469
"""2011-03""",1321,974,595500.76,869,563,411030.13,1.543517,69.022604
"""2011-04""",1149,856,469200.361,849,587,357273.97,1.446337,76.145289
"""2011-05""",1555,1056,678594.56,1271,809,566039.71,1.571075,83.413535
"""2011-06""",1393,991,661213.69,1151,773,570140.0,1.489004,86.226285
"""2011-07""",1331,949,600091.011,1143,781,532074.68,1.463508,88.665664
"""2011-08""",1280,935,645343.9,1111,778,568263.68,1.428021,88.055947
"""2011-09""",1755,1266,952838.382,1456,996,806684.471,1.461847,84.661207


### North Star propuesta
**Compras válidas de clientes recurrentes por mes.**

Justifique con los cuatro criterios: valor para el cliente, vínculo con valor empresarial, capacidad de influencia del equipo y descomposición en drivers.

**Respuesta:**  La métrica "compras válidas de clientes recurrentes por mes" cumple con los cuatro criterios de una North Star: (1) representa valor para el cliente al capturar la concreción de una transacción exitosa, evidenciando que el producto satisfizo su necesidad; (2) se vincula directamente con el valor empresarial sostenible, ya que la recurrencia es un indicador de fidelización y rentabilidad a largo plazo; (3) es influenciable por los equipos, pues áreas como marketing, operaciones y experiencia del cliente pueden ejecutar acciones concretas para impulsarla; y (4) se descompone matemáticamente en drivers operativos claros, específicamente en el número de clientes recurrentes activos y la frecuencia de compra de estos, lo que permite identificar con precisión dónde actuar para mejorar el resultado.


### Reto 1 — ¿Vanidad o acción?
Clasifique: productos totales, clientes activos mensuales, compras recurrentes, ingresos acumulados, tasa de cancelación y países con ventas. Para cada una indique qué decisión permite tomar.

**Respuesta:** Para clasificar estas métricas, es necesario distinguir entre las que solo miden escala y las que orientan decisiones concretas. Los productos totales y los países con ventas son métricas de vanidad; no permiten tomar decisiones operativas, ya que el volumen bruto o la dispersión geográfica no justifican acciones correctivas por sí solos. Los ingresos acumulados y los clientes activos mensuales son métricas de seguimiento; no dictan acciones inmediatas, sino que solo sirven para contextualizar el tamaño histórico del negocio o el alcance de las campañas de adquisición. Por el contrario, las compras recurrentes constituyen una métrica accionable (North Star) que permite decidir si es necesario reforzar las estrategias de fidelización y experiencia del cliente para asegurar valor a largo plazo. Finalmente, la tasa de cancelación actúa como un guardrail accionable que permite decidir cuándo activar revisiones operativas, de inventario o de logística para proteger la sostenibilidad y calidad del crecimiento.


# ☕ RECESO — 8:30 p. m. a 8:45 p. m.

## Paso 6 — Árbol de métricas

El árbol es una **hipótesis de gestión**, no una demostración causal.

**Objetivo → North Star → drivers → guardrails**

- Objetivo: incrementar valor recurrente sin deteriorar calidad.
- North Star: compras válidas de clientes recurrentes / mes.
- Driver 1: clientes recurrentes activos.
- Driver 2: frecuencia de compra por recurrente.
- Guardrail 1: tasa de cancelación.
- Guardrail 2: concentración de ingresos en Top 10 clientes.


In [ ]:
# Validacion aritmetica del primer nivel del arbol
kpi = kpi.with_columns(
    (pl.col("clientes_recurrentes") * pl.col("frecuencia_recurrente")).alias("ns_reconstruida"))
kpi = kpi.with_columns(
    (pl.col("compras_recurrentes") - pl.col("ns_reconstruida")).alias("error_reconstruccion"))
display(kpi.select(["mes", "compras_recurrentes", "clientes_recurrentes",
                    "frecuencia_recurrente", "error_reconstruccion"]))


mes,compras_recurrentes,clientes_recurrentes,frecuencia_recurrente,error_reconstruccion
str,u32,u32,f64,f64
"""2011-01""",570,370,1.540541,0.0
"""2011-02""",617,412,1.497573,0.0
"""2011-03""",869,563,1.543517,0.0
"""2011-04""",849,587,1.446337,1.1369e-13
"""2011-05""",1271,809,1.571075,0.0
"""2011-06""",1151,773,1.489004,0.0
"""2011-07""",1143,781,1.463508,0.0
"""2011-08""",1111,778,1.428021,0.0
"""2011-09""",1456,996,1.461847,0.0


## Paso 7 — Guardrail 1: tasa de cancelación


In [ ]:
base_total = (df.filter(pl.col("InvoiceDate").is_not_null()
                        & pl.col("CustomerID").is_not_null()
                        & (pl.col("InvoiceDate") >= INICIO)
                        & (pl.col("InvoiceDate") < FIN))
                .with_columns(pl.col("InvoiceDate").dt.strftime("%Y-%m").alias("mes")))

inv_total = base_total.select(["mes", "InvoiceNo", "CustomerID", "es_cancelacion"]).unique()
cancel = (inv_total.group_by("mes")
          .agg([pl.col("InvoiceNo").n_unique().alias("facturas_total"),
                pl.col("es_cancelacion").sum().alias("facturas_canceladas")])
          .with_columns((100 * pl.col("facturas_canceladas") / pl.col("facturas_total"))
                        .alias("tasa_cancelacion_pct"))
          .sort("mes"))
display(cancel)


mes,facturas_total,facturas_canceladas,tasa_cancelacion_pct
str,u32,u32,f64
"""2011-01""",1236,249,20.145631
"""2011-02""",1202,204,16.971714
"""2011-03""",1619,298,18.406424
"""2011-04""",1384,235,16.979769
"""2011-05""",1849,294,15.900487
"""2011-06""",1707,314,18.394845
"""2011-07""",1593,262,16.446955
"""2011-08""",1544,263,17.033679
"""2011-09""",2078,322,15.495669


## Paso 8 — Guardrail 2: concentración Top 10 clientes


In [ ]:
cliente_mes = (facturas.group_by(["mes", "CustomerID"])
               .agg(pl.col("importe_factura").sum().alias("ingreso_cliente")))

ordenado = cliente_mes.sort(["mes", "ingreso_cliente"], descending=[False, True])
ordenado = ordenado.with_columns((pl.int_range(pl.len()).over("mes") + 1).alias("ranking"))

conc = (ordenado.group_by("mes")
        .agg([pl.col("ingreso_cliente").sum().alias("ingreso_mes"),
              pl.col("ingreso_cliente").filter(pl.col("ranking") <= 10)
                .sum().alias("ingreso_top10")])
        .with_columns((100 * pl.col("ingreso_top10") / pl.col("ingreso_mes"))
                      .alias("concentracion_top10_pct"))
        .sort("mes"))

display(conc)


mes,ingreso_mes,ingreso_top10,concentracion_top10_pct
str,f64,f64,f64
"""2011-01""",569445.04,193292.89,33.944082
"""2011-02""",447137.35,89670.06,20.054254
"""2011-03""",595500.76,113585.39,19.073929
"""2011-04""",469200.361,73375.7,15.638458
"""2011-05""",678594.56,128119.53,18.880129
"""2011-06""",661213.69,193184.16,29.2166
"""2011-07""",600091.011,121223.38,20.200833
"""2011-08""",645343.9,160995.32,24.947213
"""2011-09""",952838.382,233643.58,24.520799


In [ ]:
kpi = (kpi
       .join(cancel.select(["mes", "tasa_cancelacion_pct"]), on="mes", how="left")
       .join(conc.select(["mes", "concentracion_top10_pct"]), on="mes", how="left")
       .sort("mes"))

# Variacion mensual de la North Star y de sus drivers
kpi = kpi.with_columns([
    (100 * (pl.col("compras_recurrentes") / pl.col("compras_recurrentes").shift(1) - 1)).alias("var_ns_pct"),
    (100 * (pl.col("clientes_recurrentes") / pl.col("clientes_recurrentes").shift(1) - 1)).alias("var_clientes_rec_pct"),
    (100 * (pl.col("frecuencia_recurrente") / pl.col("frecuencia_recurrente").shift(1) - 1)).alias("var_frecuencia_pct"),
])
display(kpi)


mes,facturas_validas,clientes_activos,ingresos,compras_recurrentes,clientes_recurrentes,ingresos_recurrentes,frecuencia_recurrente,participacion_ingreso_recurrente_pct,ns_reconstruida,error_reconstruccion,tasa_cancelacion_pct,concentracion_top10_pct,var_ns_pct,var_clientes_rec_pct,var_frecuencia_pct
str,u32,u32,f64,u32,u32,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64
"""2011-01""",987,741,569445.04,570,370,296614.01,1.540541,52.088259,570.0,0.0,20.145631,33.944082,null,null,null
"""2011-02""",997,758,447137.35,617,412,299937.36,1.497573,67.079469,617.0,0.0,16.971714,20.054254,8.245614,11.351351,-2.789133
"""2011-03""",1321,974,595500.76,869,563,411030.13,1.543517,69.022604,869.0,0.0,18.406424,19.073929,40.842788,36.650485,3.067901
"""2011-04""",1149,856,469200.361,849,587,357273.97,1.446337,76.145289,849.0,1.1369e-13,16.979769,15.638458,-2.301496,4.262877,-6.295983
"""2011-05""",1555,1056,678594.56,1271,809,566039.71,1.571075,83.413535,1271.0,0.0,15.900487,18.880129,49.705536,37.819421,8.624412
"""2011-06""",1393,991,661213.69,1151,773,570140.0,1.489004,86.226285,1151.0,0.0,18.394845,29.2166,-9.441385,-4.449938,-5.223907
"""2011-07""",1331,949,600091.011,1143,781,532074.68,1.463508,88.665664,1143.0,0.0,16.446955,20.200833,-0.695048,1.034929,-1.712256
"""2011-08""",1280,935,645343.9,1111,778,568263.68,1.428021,88.055947,1111.0,0.0,17.033679,24.947213,-2.79965,-0.384123,-2.424841
"""2011-09""",1755,1266,952838.382,1456,996,806684.471,1.461847,84.661207,1456.0,0.0,15.495669,24.520799,31.053105,28.020566,2.368791


### Pregunta 2
Identifique el mes con mayor tasa de cancelación y el mes con mayor concentración Top 10. ¿Qué riesgo representa cada uno?

**Respuesta:**  Según los resultados del análisis, enero de 2011 registra tanto la mayor tasa de cancelación (20.15 %) como la mayor concentración de ingresos en el Top 10 de clientes (33.94 %). La alta tasa de cancelación representa un riesgo operativo y de calidad, ya que indica posibles ineficiencias en el proceso de venta o logística que generan costos sin producir ingresos reales y pueden erosionar la confianza del cliente. Por otro lado, la elevada concentración del Top 10 evidencia un riesgo de dependencia financiera, donde la pérdida o reducción de compras de uno o pocos clientes clave comprometería de manera inmediata y severa la estabilidad de los ingresos del negocio.


## Paso 9 — Comparación reproducible con Polars y DuckDB


In [ ]:
ranking_caida = (kpi
                 .select(["mes", "compras_recurrentes", "var_ns_pct",
                          "tasa_cancelacion_pct", "concentracion_top10_pct"])
                 .sort("var_ns_pct"))
ranking_caida


mes,compras_recurrentes,var_ns_pct,tasa_cancelacion_pct,concentracion_top10_pct
str,u32,f64,f64,f64
"""2011-01""",570,null,20.145631,33.944082
"""2011-06""",1151,-9.441385,18.394845,29.2166
"""2011-08""",1111,-2.79965,17.033679,24.947213
"""2011-04""",849,-2.301496,16.979769,15.638458
"""2011-07""",1143,-0.695048,16.446955,20.200833
"""2011-10""",1571,7.898352,14.759169,21.768432
"""2011-02""",617,8.245614,16.971714,20.054254
"""2011-09""",1456,31.053105,15.495669,24.520799
"""2011-03""",869,40.842788,18.406424,19.073929


In [ ]:
# DuckDB consulta el DataFrame de Polars directamente, sin copias intermedias.
consulta = duckdb.sql("""
SELECT mes,
       compras_recurrentes,
       ROUND(var_ns_pct, 1) AS var_ns_pct,
       ROUND(tasa_cancelacion_pct, 1) AS cancelacion_pct,
       ROUND(concentracion_top10_pct, 1) AS concentracion_top10_pct
FROM kpi
ORDER BY var_ns_pct ASC NULLS LAST
""").pl()
display(consulta)


mes,compras_recurrentes,var_ns_pct,cancelacion_pct,concentracion_top10_pct
str,u32,f64,f64,f64
"""2011-06""",1151,-9.4,18.4,29.2
"""2011-08""",1111,-2.8,17.0,24.9
"""2011-04""",849,-2.3,17.0,15.6
"""2011-07""",1143,-0.7,16.4,20.2
"""2011-10""",1571,7.9,14.8,21.8
"""2011-02""",617,8.2,17.0,20.1
"""2011-09""",1456,31.1,15.5,24.5
"""2011-03""",869,40.8,18.4,19.1
"""2011-11""",2334,48.6,13.9,15.7


## Paso 10 — Tablero de decisión


In [ ]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=kpi["mes"].to_list(), y=kpi["compras_recurrentes"].to_list(),
                         mode="lines+markers", name="North Star"))

# El tablero incorpora el driver principal junto al resultado: un tablero que
# muestra la North Star sin su driver no permite decidir donde actuar.
fig.add_trace(go.Scatter(x=kpi["mes"].to_list(), y=kpi["clientes_recurrentes"].to_list(),
                         mode="lines+markers", name="Clientes recurrentes (driver)"))

fig.update_layout(title="North Star y driver principal - 2011", xaxis_title="Mes",
                  yaxis_title="Cantidad", hovermode="x unified")
fig.show()


In [ ]:
fig2 = px.line(kpi.to_pandas(), x="mes", y=["tasa_cancelacion_pct", "concentracion_top10_pct"],
               markers=True, title="Guardrails observados - cancelacion y concentracion")
fig2.update_layout(yaxis_title="Porcentaje (%)", xaxis_title="Mes", legend_title_text="Guardrail")
fig2.show()


In [ ]:
fig3 = px.line(kpi.to_pandas(), x="mes", y=["participacion_ingreso_recurrente_pct"],
               markers=True, title="Participacion del ingreso proveniente de compras recurrentes")
fig3.update_layout(yaxis_title="Porcentaje (%)", xaxis_title="Mes")
fig3.show()


### Interpretación obligatoria
No basta con decir “la línea bajó”. Responda:
1. ¿Cuánto cambió y en qué periodo?
2. ¿Cómo se comportó el driver principal?
3. ¿Algún guardrail empeoró al mismo tiempo?
4. ¿Qué puede afirmarse como evidencia y qué es solo una inferencia?
5. ¿Quién debería decidir y qué acción ejecutaría?

**Respuesta:**  En junio de 2011, la North Star (compras recurrentes) experimentó su mayor caída mensual con un descenso del 9,44 % respecto al periodo anterior. Este resultado fue explicado matemáticamente por la disminución simultánea de sus dos drivers: los clientes recurrentes activos bajaron un 4,45 % y la frecuencia de compra un 5,22 %. Al mismo tiempo, ambos guardrails empeoraron, ya que la tasa de cancelación aumentó al 18,39 % y la concentración de ingresos en el Top 10 de clientes se elevó al 29,22 %. Como evidencia, los datos confirman un deterioro generalizado en el volumen y la calidad de las transacciones; sin embargo, inferir la causa raíz (como fallas logísticas o de producto) requiere un análisis cualitativo adicional, ya que los datos solo muestran correlación, no causalidad. Ante este escenario, el responsable de Operaciones o Experiencia del Cliente debería priorizar una auditoría de las facturas canceladas y de la dependencia de los clientes principales de ese mes, deteniendo temporalmente nuevas inversiones en adquisición hasta estabilizar la calidad del servicio.


## Reto 2 — Diagnóstico sin confundir correlación con causalidad


In [ ]:
# Mes con mayor caida porcentual de la North Star.
# Se excluye el mes de calentamiento y el primero comparado contra el.
peor = (kpi.filter((pl.col("mes") > MES_CALENTAMIENTO) & pl.col("var_ns_pct").is_not_null())
        .sort("var_ns_pct")
        .head(1))
print("Mayor caida mensual de la North Star")
display(peor.select(["mes", "var_ns_pct", "var_clientes_rec_pct", "var_frecuencia_pct",
                     "tasa_cancelacion_pct", "concentracion_top10_pct"]))


Mayor caida mensual de la North Star


mes,var_ns_pct,var_clientes_rec_pct,var_frecuencia_pct,tasa_cancelacion_pct,concentracion_top10_pct
str,f64,f64,f64,f64,f64
"""2011-06""",-9.441385,-4.449938,-5.223907,18.394845,29.2166


A partir de la salida anterior, escriba una conclusión en tres capas:

- **Evidencia:** lo que muestran los datos.
- **Inferencia:** explicación plausible, sin afirmar causalidad.
- **Decisión:** acción concreta que debería evaluar el responsable.

**Respuesta:**

* Evidencia: En junio de 2011, la North Star (compras recurrentes) registró su mayor caída mensual con un descenso del 9,44 %, explicado matemáticamente por la disminución simultánea de sus dos drivers: clientes recurrentes (-4,45 %) y frecuencia de compra (-5,22 %). Adicionalmente, los guardrails se deterioraron en el mismo periodo, con una tasa de cancelación del 18,39 % y una concentración de ingresos en el Top 10 del 29,22 %.
* Inferencia: La contracción simultánea de la base de clientes leales y su frecuencia, acompañada de un repunte en las cancelaciones, sugiere un deterioro sistémico en la calidad del servicio o en la operación logística durante ese mes específico, lo cual erosionó la confianza y la experiencia del segmento más valioso del negocio.
* Decisión: El responsable de Operaciones o Experiencia del Cliente debe priorizar una auditoría de las facturas canceladas de junio de 2011 y realizar un sondeo directo a los clientes del Top 10 para identificar la causa raíz de la fricción, pausando temporalmente las inversiones en adquisición hasta estabilizar la calidad del servicio.


## Diccionario de KPI — completar

| KPI | Fórmula / unidad | Fuente / frecuencia | Responsable | Meta / alerta | Acción |
|---|---|---|---|---|---|
| Compras recurrentes/mes |Suma de facturas válidas donde el cliente tiene ≥ 2 compras históricas. / Cantidad de facturas. | UCI / mensual | Gerente de Experiencia del Cliente y Fidelización | Meta: Crecimiento mensual > 5%. Alerta: Caída mensual > 5% (ej. -9.4% observado en jun-2011). | Revisar campañas de retención, analizar causas de abandono de clientes leales y mejorar el servicio postventa. |
| Clientes recurrentes activos | Conteo de clientes únicos (CustomerID) con al menos dos compras válidas en el periodo. / Cantidad de clientes. | UCI / mensual | Gerente de CRM y Retención | Meta: Aumentar la base recurrente un 3% mensual. Alerta: Disminución > 4% (ej. -4.4% observado en jun-2011). | Activar programas de re-engagement, evaluar satisfacción de clientes leales y personalizar ofertas de recompra. |
| Frecuencia recurrente | (Compras recurrentes / Clientes recurrentes activos). / Compras por cliente. | UCI / mensual | Gerente de Producto y Catálogo | Meta: Mantener o superar 1.55 compras por cliente. Alerta: Caer por debajo de 1.45 (ej. 1.42 observado en ago-2011). | Optimizar el catálogo, implementar estrategias de venta cruzada (cross-selling) y revisar disponibilidad de stock. |
| Tasa de cancelación | (Facturas canceladas / Total de facturas emitidas en el mes) * 100. / Porcentaje (%). | UCI / mensual | Gerente de Operaciones y Logística | Meta: Mantener por debajo del 15%. Alerta: Superar el 18% (ej. 20.1% observado en ene-2011). | Auditar procesos de facturación y logística, revisar calidad de inventario y capacitar al equipo de atención al cliente. |
| Concentración Top 10 | (Ingreso de los 10 clientes con mayor facturación / Ingreso total del mes) * 100. / Porcentaje (%). | UCI / mensual | Director Comercial / Gerente de Ventas | Meta: Mantener por debajo del 20%. Alerta: Superar el 25% (ej. 33.9% observado en ene-2011). | Diversificar la cartera de clientes, intensificar la adquisición de clientes medianos y asegurar contratos con los principales. |

> Las metas y alertas que proponga son supuestos académicos de gestión, no metas oficiales de la empresa.


---

## Reto de aplicación y retroalimentación

En esta sección se aplicarán los procedimientos desarrollados durante la sesión a nuevas situaciones de análisis. Cada ejercicio requiere modificar, completar o construir código a partir de las tablas ya procesadas. Posteriormente, los resultados obtenidos deberán interpretarse brevemente desde una perspectiva empresarial. El propósito es comprobar la comprensión de las técnicas utilizadas y fortalecer la capacidad de adaptar el análisis ante nuevas preguntas de negocio.

**Indicaciones generales.** Los ejercicios operan sobre `kpi_reto`, copia de trabajo de la tabla de KPI, y sobre `facturas`, ya construida en la Actividad 2. Los datos proceden íntegramente del repositorio UCI; no corresponde generar ni sustituir valores en ningún caso. Cada respuesta escrita no debe exceder cuatro líneas.

**Duración en sesión:** 25 minutos. Los ejercicios que no concluyan se completan como avance del entregable.


In [ ]:
# Copia de trabajo para la sección de retos.
kpi_reto = kpi.clone()
print("Copia de trabajo creada:", kpi_reto.shape)
print("Columnas disponibles:", kpi_reto.columns)
display(kpi_reto.head())


Copia de trabajo creada: (11, 16)
Columnas disponibles: ['mes', 'facturas_validas', 'clientes_activos', 'ingresos', 'compras_recurrentes', 'clientes_recurrentes', 'ingresos_recurrentes', 'frecuencia_recurrente', 'participacion_ingreso_recurrente_pct', 'ns_reconstruida', 'error_reconstruccion', 'tasa_cancelacion_pct', 'concentracion_top10_pct', 'var_ns_pct', 'var_clientes_rec_pct', 'var_frecuencia_pct']


mes,facturas_validas,clientes_activos,ingresos,compras_recurrentes,clientes_recurrentes,ingresos_recurrentes,frecuencia_recurrente,participacion_ingreso_recurrente_pct,ns_reconstruida,error_reconstruccion,tasa_cancelacion_pct,concentracion_top10_pct,var_ns_pct,var_clientes_rec_pct,var_frecuencia_pct
str,u32,u32,f64,u32,u32,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64
"""2011-01""",987,741,569445.04,570,370,296614.01,1.540541,52.088259,570.0,0.0,20.145631,33.944082,null,null,null
"""2011-02""",997,758,447137.35,617,412,299937.36,1.497573,67.079469,617.0,0.0,16.971714,20.054254,8.245614,11.351351,-2.789133
"""2011-03""",1321,974,595500.76,869,563,411030.13,1.543517,69.022604,869.0,0.0,18.406424,19.073929,40.842788,36.650485,3.067901
"""2011-04""",1149,856,469200.361,849,587,357273.97,1.446337,76.145289,849.0,1.1369e-13,16.979769,15.638458,-2.301496,4.262877,-6.295983
"""2011-05""",1555,1056,678594.56,1271,809,566039.71,1.571075,83.413535,1271.0,0.0,15.900487,18.880129,49.705536,37.819421,8.624412


### Ejercicio 1 — Construcción de una tasa de recurrencia mensual

La North Star del laboratorio se expresa en cantidad de compras recurrentes, magnitud absoluta que crece cuando aumenta la actividad total del negocio. Una magnitud absoluta, sin embargo, no permite distinguir si la recurrencia mejora o si simplemente hay más transacciones de cualquier tipo.

La tasa de recurrencia corrige esa limitación: expresa qué proporción de las facturas del mes corresponde a clientes que ya habían comprado con anterioridad. Se trata de una magnitud relativa y, por tanto, comparable entre meses de distinto volumen.

**Se solicita:**

1. Calcular, para cada mes, la cantidad total de facturas y la cantidad de facturas recurrentes a partir de `facturas`.
2. Construir la tasa de recurrencia mensual expresada en porcentaje e incorporarla a `kpi_reto`.
3. Identificar el mes con la tasa más alta y el mes con la tasa más baja.
4. Comparar el ordenamiento por tasa de recurrencia con el ordenamiento por compras recurrentes.

**Tiempo estimado:** 5 minutos.


In [ ]:
# ---------------------------------------------------------------------------
# CELDA DE TRABAJO — Ejercicio 1
# Desarrolle aquí su código.
# ---------------------------------------------------------------------------

# 1. Calcular, para cada mes, la cantidad total de facturas y la cantidad de facturas recurrentes a partir de 'facturas'
agg_facturas = (
    facturas.group_by("mes")
    .agg([
        pl.col("InvoiceNo").n_unique().alias("total_facturas"),
        pl.col("InvoiceNo").filter(pl.col("es_recurrente")).n_unique().alias("facturas_recurrentes")
    ])
    .sort("mes")
)

# 2. Construir la tasa de recurrencia mensual en porcentaje e incorporarla a kpi_reto
agg_facturas = agg_facturas.with_columns(
    (100 * pl.col("facturas_recurrentes") / pl.col("total_facturas")).alias("tasa_recurrencia_pct")
)

kpi_reto = kpi_reto.join(
    agg_facturas.select(["mes", "tasa_recurrencia_pct"]),
    on="mes",
    how="left"
)

# 3. Identificar el mes con la tasa más alta y el mes con la tasa más baja
print("--- Mes con tasa de recurrencia más alta ---")
display(kpi_reto.sort("tasa_recurrencia_pct", descending=True).select(["mes", "tasa_recurrencia_pct"]).head(1))

print("--- Mes con tasa de recurrencia más baja ---")
display(kpi_reto.sort("tasa_recurrencia_pct", descending=False).select(["mes", "tasa_recurrencia_pct"]).head(1))

# 4. Comparar el ordenamiento por tasa de recurrencia con el ordenamiento por compras recurrentes
print("--- Comparación de ordenamientos (ordenado de mayor a menor por tasa de recurrencia) ---")
display(kpi_reto.select(["mes", "compras_recurrentes", "tasa_recurrencia_pct"]).sort("tasa_recurrencia_pct", descending=True))

--- Mes con tasa de recurrencia más alta ---


mes,tasa_recurrencia_pct
str,f64
"""2011-11""",87.843432


--- Mes con tasa de recurrencia más baja ---


mes,tasa_recurrencia_pct
str,f64
"""2011-01""",57.75076


--- Comparación de ordenamientos (ordenado de mayor a menor por tasa de recurrencia) ---


mes,compras_recurrentes,tasa_recurrencia_pct
str,u32,f64
"""2011-11""",2334,87.843432
"""2011-08""",1111,86.796875
"""2011-07""",1143,85.875282
"""2011-09""",1456,82.962963
"""2011-06""",1151,82.627423
"""2011-05""",1271,81.736334
"""2011-10""",1571,81.441161
"""2011-04""",849,73.890339
"""2011-03""",869,65.783497


**Pregunta 3.** ¿Coincide el mes con mayor cantidad de compras recurrentes con el mes de mayor tasa de recurrencia? Explique qué implicancia tiene esa diferencia para la medición del negocio.

**Respuesta (máximo cuatro líneas):**  Sí, ambos indicadores coinciden en su máximo durante noviembre de 2011. No obstante, el ordenamiento general difiere (ej. agosto tiene alta tasa pero bajo volumen, y octubre alto volumen pero menor tasa). Esta discrepancia demuestra que las magnitudes absolutas pueden ocultar la verdadera dependencia del negocio hacia clientes leales, por lo que la tasa relativa es indispensable para evaluar la salud real de la recurrencia.

---


### Ejercicio 2 — Incorporación del ticket promedio como driver económico

El árbol de métricas descompuso la North Star en clientes recurrentes y frecuencia de compra. Ninguno de esos dos drivers recoge el valor económico de cada transacción: dos meses con idéntica cantidad de compras recurrentes pueden presentar ingresos muy distintos si el importe promedio de la factura cambia.

El ticket promedio de la factura recurrente completa esa descripción y permite establecer si el crecimiento observado proviene de mayor actividad o de mayor valor por transacción.

**Se solicita:**

1. Calcular, a partir de `facturas`, el importe promedio de las facturas recurrentes de cada mes.
2. Incorporar el resultado a `kpi_reto` en la columna `ticket_promedio_recurrente`.
3. Calcular su variación mensual porcentual, siguiendo el mismo procedimiento empleado para los demás drivers.
4. Contrastar los meses de mayor caída de la North Star con el comportamiento del ticket promedio en esos mismos meses.

**Tiempo estimado:** 5 minutos.


In [ ]:
# ---------------------------------------------------------------------------
# CELDA DE TRABAJO — Ejercicio 2
# Desarrolle aquí su código.
# ---------------------------------------------------------------------------

# 1. Calcular, a partir de 'facturas', el importe promedio de las facturas recurrentes de cada mes
ticket_promedio = (
    facturas.filter(pl.col("es_recurrente"))
    .group_by("mes")
    .agg(pl.col("importe_factura").mean().alias("ticket_promedio_recurrente"))
)

# 2. Incorporar el resultado a kpi_reto
kpi_reto = kpi_reto.join(ticket_promedio, on="mes", how="left")

# 3. Calcular su variación mensual porcentual (siguiendo el procedimiento de los otros drivers)
kpi_reto = kpi_reto.with_columns(
    (100 * (pl.col("ticket_promedio_recurrente") / pl.col("ticket_promedio_recurrente").shift(1) - 1))
    .round(2)
    .alias("var_ticket_promedio_pct")
)

# 4. Contrastar los meses de mayor caída de la North Star con el comportamiento del ticket promedio
print("--- Meses con mayor caída de la North Star y comportamiento del ticket promedio ---")
display(
    kpi_reto.select(["mes", "var_ns_pct", "ticket_promedio_recurrente", "var_ticket_promedio_pct"])
    .sort("var_ns_pct")
    .head(3)
)

--- Meses con mayor caída de la North Star y comportamiento del ticket promedio ---


mes,var_ns_pct,ticket_promedio_recurrente,var_ticket_promedio_pct
str,f64,f64,f64
"""2011-01""",null,520.375456,null
"""2011-06""",-9.441385,495.34318,11.23
"""2011-08""",-2.79965,511.488461,9.88


**Pregunta 4.** En el mes de mayor caída de la North Star, ¿el ticket promedio recurrente aumentó o disminuyó? ¿Qué sugiere ese comportamiento conjunto?

**Respuesta (máximo cuatro líneas):**  En junio de 2011, mes con la mayor caída de la North Star, el ticket promedio recurrente aumentó un 11,23 %. Este comportamiento sugiere que la contracción del negocio fue impulsada por una pérdida de volumen (menos clientes y menor frecuencia) y no por una disminución en el valor económico de las transacciones. Esto evidencia que las fricciones operativas redujeron la cantidad de compras exitosas, pero no afectaron la capacidad de gasto de los clientes que sí lograron concretar su transacción.

---


### Ejercicio 3 — Definición de un guardrail de dependencia del cliente principal

La Actividad 2 incorporó un guardrail de concentración sobre los diez principales clientes. Ese umbral describe la dependencia agregada del negocio, pero no revela si dicha concentración se explica por un único cliente de gran tamaño, situación que constituye un riesgo operativo de naturaleza distinta.

Un guardrail de dependencia del cliente principal mide qué proporción del ingreso mensual corresponde al cliente de mayor facturación. Su finalidad no es optimizarse, sino advertir cuando la concentración alcanza un nivel que compromete la continuidad del negocio.

**Se solicita:**

1. Calcular, para cada mes, el ingreso del cliente de mayor facturación. La tabla `ordenado` ya dispone de la columna `ranking` calculada dentro de cada mes.
2. Expresar ese ingreso como porcentaje del ingreso total del mes e incorporarlo a `kpi_reto`.
3. Declarar explícitamente un umbral de alerta como criterio pedagógico del ejercicio, no como norma sectorial.
4. Identificar los meses que superan dicho umbral.

**Tiempo estimado:** 5 minutos.


In [ ]:
# ---------------------------------------------------------------------------
# CELDA DE TRABAJO — Ejercicio 3
# Desarrolle aquí su código.
# ---------------------------------------------------------------------------

# 1. Calcular el ingreso del cliente de mayor facturación (ranking == 1) por mes
top1_cliente = (
    ordenado.filter(pl.col("ranking") == 1)
    .select(["mes", pl.col("ingreso_cliente").alias("ingreso_top1_cliente")])
)

# 2. Incorporar el resultado a kpi_reto y calcular el porcentaje sobre el ingreso total del mes
kpi_reto = kpi_reto.join(top1_cliente, on="mes", how="left")
kpi_reto = kpi_reto.with_columns(
    (100 * pl.col("ingreso_top1_cliente") / pl.col("ingresos")).alias("dependencia_top1_pct")
)

# 3. Declarar explícitamente un umbral de alerta (criterio pedagógico ajustado)
UMBRAL_DEPENDENCIA = 10.0  # Ajustado a 10% para capturar meses con mayor concentración individual

# 4. Identificar los meses que superan dicho umbral
meses_riesgo = (
    kpi_reto.filter(pl.col("dependencia_top1_pct") > UMBRAL_DEPENDENCIA)
    .select(["mes", "ingreso_top1_cliente", "ingresos", "dependencia_top1_pct"])
    .sort("dependencia_top1_pct", descending=True)
)

print(f"Meses que superan el umbral de dependencia del {UMBRAL_DEPENDENCIA}%:")
display(meses_riesgo)

Meses que superan el umbral de dependencia del 10.0%:


mes,ingreso_top1_cliente,ingresos,dependencia_top1_pct
str,f64,f64,f64
"""2011-01""",77183.6,569445.04,13.554179


**Pregunta 5.** ¿En cuántos meses el cliente principal supera el umbral declarado y qué decisión empresarial justificaría ese resultado?

**Respuesta (máximo cuatro líneas):**  Solo un mes (enero de 2011) supera el umbral del 10 %, con una dependencia del 13,55 %. Esto evidencia que, en general, el negocio no presenta un riesgo crítico de concentración en un único cliente. No obstante, justifica la decisión de implementar estrategias proactivas de diversificación de la cartera y monitoreo de los clientes de mayor volumen para blindar la estabilidad de los ingresos ante posibles bajas en su demanda.

---


### Ejercicio 4 — Consulta de meses saludables con condiciones múltiples en DuckDB

En la Actividad 2 se empleó DuckDB para ordenar los meses según la variación de la North Star. Una consulta orientada a la decisión, sin embargo, no se limita a ordenar: delimita el subconjunto de periodos que satisfacen simultáneamente el objetivo de crecimiento y las restricciones fijadas por los guardrails.

Un mes de crecimiento acompañado de un deterioro en la tasa de cancelación no constituye un mes saludable. Esta consulta materializa esa distinción en una regla reproducible.

**Se solicita:**

1. Construir sobre `kpi_reto` una consulta SQL que incluya `SELECT`, `WHERE`, condiciones múltiples enlazadas con `AND` y `ORDER BY`.
2. Establecer como criterios una variación de la North Star positiva y una tasa de cancelación inferior al promedio del periodo.
3. Ordenar el resultado por variación de la North Star de mayor a menor.
4. Determinar cuántos meses satisfacen ambos criterios.

**Tiempo estimado:** 5 minutos.


In [ ]:
# ---------------------------------------------------------------------------
# CELDA DE TRABAJO — Ejercicio 4
# Desarrolle aquí su código.
# ---------------------------------------------------------------------------

# Consulta SQL en DuckDB para identificar meses "saludables"
# Criterios: variación de North Star positiva Y tasa de cancelación inferior al promedio del periodo
consulta_meses_saludables = duckdb.sql("""
    SELECT
        mes,
        ROUND(var_ns_pct, 2) AS var_ns_pct,
        ROUND(tasa_cancelacion_pct, 2) AS tasa_cancelacion_pct
    FROM kpi_reto
    WHERE var_ns_pct > 0
      AND tasa_cancelacion_pct < (
          SELECT AVG(tasa_cancelacion_pct)
          FROM kpi_reto
          WHERE var_ns_pct IS NOT NULL
      )
    ORDER BY var_ns_pct DESC
""").pl()

print("--- Meses saludables (crecimiento positivo y cancelación bajo el promedio) ---")
display(consulta_meses_saludables)

print(f"\nTotal de meses que satisfacen ambos criterios: {consulta_meses_saludables.height}")

--- Meses saludables (crecimiento positivo y cancelación bajo el promedio) ---


mes,var_ns_pct,tasa_cancelacion_pct
str,f64,f64
"""2011-05""",49.71,15.9
"""2011-11""",48.57,13.87
"""2011-09""",31.05,15.5
"""2011-10""",7.9,14.76



Total de meses que satisfacen ambos criterios: 4


**Pregunta 6.** ¿Cuántos meses del periodo pueden calificarse como saludables según los criterios establecidos y qué sugiere esa proporción sobre la solidez del crecimiento?

**Respuesta (máximo cuatro líneas):**  Cuatro meses del periodo (mayo, septiembre, octubre y noviembre) cumplen con los criterios de crecimiento positivo y calidad operativa. Esta proporción, que representa aproximadamente un tercio del periodo analizado, sugiere que el crecimiento del negocio no es consistentemente sólido, ya que en la mayoría de los meses la expansión se acompañó de un deterioro en la tasa de cancelación, evidenciando fricciones operativas recurrentes durante la escalada de ventas.

---


### Ejercicio 5 — Representación de la relación entre el driver y la North Star

El tablero construido en la Actividad 2 presenta la North Star y su driver como series temporales paralelas. Esa disposición permite observar la evolución de ambas magnitudes, pero no muestra con claridad si sus variaciones se corresponden entre sí.

Un gráfico de dispersión que enfrente la variación del driver con la variación de la North Star hace visible esa correspondencia: los puntos alineados sobre una tendencia ascendente indican que el driver acompaña el movimiento del resultado, mientras que los puntos dispersos advierten que otros factores intervienen. Corresponde recordar que la correspondencia observada describe una asociación y no acredita una relación causal.

**Se solicita:**

1. Construir con Plotly un gráfico de dispersión que sitúe la variación porcentual de clientes recurrentes en el eje horizontal y la variación porcentual de la North Star en el eje vertical.
2. Identificar cada punto con el mes correspondiente.
3. Rotular los ejes y titular el gráfico de modo que resulte interpretable sin recurrir al código.
4. Excluir de manera explícita el mes de calentamiento, que carece de variación calculable.

**Tiempo estimado:** 5 minutos.


In [ ]:
# ---------------------------------------------------------------------------
# CELDA DE TRABAJO — Ejercicio 5
# Desarrolle aquí su código.
# ---------------------------------------------------------------------------

import plotly.express as px

# 1. Excluir explícitamente el mes de calentamiento (2011-01), que carece de variación calculable
datos_grafico = kpi_reto.filter(pl.col("mes") != "2011-01").to_pandas()

# 2. Construir el gráfico de dispersión con Plotly
fig = px.scatter(
    datos_grafico,
    x="var_clientes_rec_pct",
    y="var_ns_pct",
    text="mes",               # Identifica cada punto con el mes correspondiente
    hover_name="mes",         # Muestra el mes al pasar el cursor
    title="Relación entre la variación del driver (clientes recurrentes) y la North Star",
    labels={
        "var_clientes_rec_pct": "Variación mensual de clientes recurrentes (%)",
        "var_ns_pct": "Variación mensual de la North Star (%)"
    }
)

# 3. Ajustes de diseño para mejorar la legibilidad sin recurrir al código
fig.update_traces(textposition="top center", marker=dict(size=10))
fig.update_layout(
    xaxis=dict(zeroline=True, zerolinewidth=1, zerolinecolor="black"),
    yaxis=dict(zeroline=True, zerolinewidth=1, zerolinecolor="black"),
    hovermode="closest",
    width=800,
    height=500
)

fig.show()

**Pregunta 7.** ¿La variación del driver acompaña la variación de la North Star? Indique un mes que se aparte de esa correspondencia y proponga una explicación verificable.

**Respuesta (máximo cuatro líneas):**  Sí, la variación del driver acompaña claramente la variación de la North Star, como lo evidencia la tendencia ascendente en el gráfico de dispersión. El mes de abril de 2011 se aparta de esta correspondencia: los clientes recurrentes aumentaron 4,26 %, pero la North Star cayó 2,30 %. La explicación verificable es que la frecuencia de compra de los clientes recurrentes disminuyó 6,30 % en ese mismo mes, compensando negativamente el crecimiento de la base de clientes.


---


## Informe ejecutivo breve

1. **Objetivo estratégico:**  

    Transformar las transacciones históricas en un sistema de métricas que distinga crecimiento útil de crecimiento aparente, identificando el valor recurrente del cliente para orientar decisiones empresariales concretas.

2. **North Star y justificación:**  

    Compras válidas de clientes recurrentes por mes. Representa valor para el cliente (transacción completada), se vincula con sostenibilidad empresarial (recurrencia), es influenciable por los equipos y se descompone en drivers medibles (clientes recurrentes y frecuencia).

3. **Hallazgo cuantitativo 1:**  

    Junio de 2011 registró la mayor caída de la North Star (-9,44 %), explicada por la disminución simultánea de clientes recurrentes (-4,45 %) y frecuencia de compra (-5,22 %), acompañada de un deterioro en ambos guardrails.

4. **Hallazgo cuantitativo 2:**  

    Enero de 2011 presentó la mayor tasa de cancelación (20,15 %) y la mayor concentración del Top 10 (33,94 %), evidenciando riesgos operativos y de dependencia de clientes en el inicio del periodo.

5. **Hallazgo cuantitativo 3:**  

    Solo 4 de 11 meses (mayo, septiembre, octubre y noviembre) calificaron como "saludables" al cumplir crecimiento positivo y tasa de cancelación bajo el promedio, sugiriendo que el crecimiento no fue consistentemente sólido.

6. **Driver prioritario:**  

    Clientes recurrentes activos, ya que su variación explica la mayor parte del movimiento de la North Star y representa la base de clientes leales que genera valor sostenible.

7. **Guardrails:**  

    Tasa de cancelación (protege la calidad operativa) y concentración de ingresos en Top 10 clientes (protege contra dependencia excesiva). Ambos deben monitorearse para evitar que la optimización de la North Star sacrifique calidad o estabilidad.

8. **Decisión empresarial recomendada:**  

    El responsable de Operaciones debe priorizar una auditoría de las facturas canceladas en meses de caída (especialmente junio 2011) y fortalecer la diversificación de la cartera de clientes, pausando inversiones en adquisición hasta estabilizar la calidad del servicio y reducir la dependencia de clientes principales.

9. **Limitación del dataset:**

    El conjunto corresponde a un solo minorista del Reino Unido durante un periodo específico (diciembre 2010 - diciembre 2011), carece de información demográfica de clientes y no permite análisis de cohortes detallado ni segmentación por comportamiento más allá de la recurrencia.


## Ticket de salida

Elija una métrica de su proyecto integrador. Explique por qué es accionable y qué decisión cambiaría si disminuyera 20 %.

**Respuesta:**  

Con la información proporcionada no es posible determinarlo con precisión, ya que el "proyecto integrador" corresponde a un trabajo personal o de equipo externo al caso "Online Retail" que hemos analizado en este laboratorio.
Para poder redactar esta respuesta, necesito que me indiques brevemente:
1. El tema o contexto de tu proyecto integrador.
2. Una métrica clave que hayas seleccionado para dicho proyecto.



## Referencias

- Chen, D. (2015). *Online Retail* [Dataset]. UCI Machine Learning Repository. https://doi.org/10.24432/C5BW33  
- Croll, A., & Yoskovitz, B. (2013). *Lean Analytics*. O’Reilly Media.  
- Parmenter, D. (2020). *Key Performance Indicators* (4th ed.). Wiley.  
- Sharda, R., Delen, D., & Turban, E. (2024). *Business Intelligence, Analytics, Data Science, and AI* (5th ed.). Pearson.
